runtime/nvqir/qpp/QppCircuitSimulator.cpp

```diff
@@ -24,63 +25,50 @@ class QppCircuitSimulator : public nvqir::CircuitSimulatorBase<double> {
   /// The QPP state representation (qpp::ket or qpp::cmat)
   StateType state;
 
-  /// Convert from little endian to big endian.
-  std::size_t bigEndian(const int n_qubits, const int bit) {
-    return n_qubits - bit - 1;
+  /// @brief Convert internal qubit index to Q++ qubit index.
+  ///
+  /// In Q++, qubits are indexed from left to right, and thus q0 is the leftmost
+  /// qubit. Internally, in CUDA Quantum, qubits are index from right to left,
+  /// hence q0 is the rightmost qubit. Example:
+  /// ```
+  ///   Q++ indices:  0  1  2  3
+  ///                |0>|0>|0>|0>
+  ///                 3  2  1  0 : CUDA Quantum indices
+  /// ```
+  std::size_t convertQubitIndex(std::size_t qubitIndex) {
+    assert(stateDimension > 0 && "The state is empty, and thus has no qubits");
+    return std::log2(stateDimension) - qubitIndex - 1;
   }
 
   /// @brief Compute the expectation value <Z...Z> over the given qubit indices.
-  /// @param qubit_indices
-  /// @return expectation
-  double
-  calculateExpectationValue(const std::vector<std::size_t> &qubit_indices) {
-    const auto hasEvenParity =
-        [](std::size_t x,
-           const std::vector<std::size_t> &in_qubitIndices) -> bool {
-      size_t count = 0;
-      for (const auto &bitIdx : in_qubitIndices) {
-        if (x & (1ULL << bitIdx)) {
-          count++;
-        }
-      }
-      return (count % 2) == 0;
+  double calculateExpectationValue(const std::vector<std::size_t> &qubits) {
+    const auto hasEvenParity = [&qubits](std::size_t x) -> bool {
+      std::size_t bitmask = 0;
+      for (auto q : qubits)
+        bitmask |= (1ULL << q);
+      return std::popcount(x & bitmask) % 2 == 0;
     };
```

Comment:
Thanks @boschmitt 👍 for addressing this in this PR. 

We should also move the `bitmask` construction out of the lambda: 
  
```suggestion
    std::size_t bitmask = 0;
    for (auto q : qubits)
      bitmask |= (1ULL << q);
    const auto hasEvenParity = [&bitmask](std::size_t x) -> bool {
      return std::popcount(x & bitmask) % 2 == 0;
    };
```